onnxruntime/core/providers/coreml/coreml_execution_provider.cc

```diff
@@ -57,7 +58,18 @@ CoreMLExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph_vie
       [&]() {
         HashValue model_hash;
         int metadef_id = metadef_id_generator_.GenerateId(graph_viewer, model_hash);
-        return MakeString(COREML, "_", model_hash, "_", metadef_id);
+        std::string user_provide_key;
+        const Graph* main_graph = &graph_viewer.GetGraph();
+        while (main_graph->IsSubgraph()) {
+          main_graph = main_graph->ParentGraph();
+        }
+        if (main_graph->GetModel().MetaData().count("CACHE_KEY") > 0) {
+          user_provide_key = graph_viewer.GetGraph().GetModel().MetaData().at("CACHE_KEY");
```

Comment:
We need to validate this with something like std::isalnum to guarantee it will be valid for use in the filesystem. I'd also suggest we enforce a maximum length to also try and avoid issues creating folders/files with this in the name.

Reply:
Yeah. we should do it.
For now, cache-key should has at most 32 characters. Or we will re-hash it.

Reply:
64 chars + null is required to store a sha256 hash as hex so maybe we could use that as the limit?

Reply:
64 comes into to my mind in the first thought, would this be too long?

Reply:
it's a maximum so the user isn't forced to use it all. Allowing a 256 bit hash seems reasonable. Max directory/file name limit is 255 chars so 64 is well under that. Max path is somewhere around 1000 chars. User also controls the cache path.

But we should be careful about how deep the directory names get for the cache files. It might make more sense to shorten things like whether static shapes were enabled, and for the model it's really only the subgraph id that matters. 

i.e. including the user hash and the model hash in a directory name (userhash_COREML_modelhash_subgraphid) shouldn't actually matter when all files for the model are stored under a top-level directory name that uses the preferred hash, as there a) should never be files from any other models in there, and b) the subgraphid should be deterministic (unless they run with a different optimization level). 

Take a look at how long the paths are in your tests and figure out a good balance between readable/safe and avoiding exceeding the max path length.

Reply:
Sure. It's quite balance now.
<user_provided_path>/cc6b2111b15dcdcf00ed6647c4430315e01616efde46ab4109f80fd1d3c46731/0_dynamic_mlprogram/model/